### PR TITLE
perf: small Table performance fixes

### PR DESCRIPTION
**Pull request by krozhkov**
* remove linear complexity from getRowId function in a compatible way
* add enhanced columns memoization to table with settings hoc
* cache css styles for cells.

### CODE REVIEW

File: src/components/Table/Table.tsx [hunk 1]
```diff
@@ -101,9 +101,8 @@ export class Table<I extends TableDataItem = Record<string, string>> extends Rea
     };
 
     // Static methods may be used by HOCs
-    static getRowId<I extends TableDataItem>(props: TableProps<I>, item: I) {
-        const {data, getRowId} = props;
-        const index = data.indexOf(item);
+    static getRowId<I extends TableDataItem>(props: TableProps<I>, item: I, index: number) {
```
Comment (amje) [inline src/components/Table/Table.tsx:104]: This is a breaking change, not sure we are ready to next major now

File: src/components/Table/Table.tsx [hunk 2]
```diff
@@ -18,8 +18,8 @@ interface TableState {
     // activeScrollElement нужен, чтобы listner на скролл таблицы не срабатывал, когда скроллится скролл-бар и наоборот
     // в противном случае таблица будет дергаться при прокрутке
     activeScrollElement: ActiveScrollElementType;
-    columnsWidth: (number | undefined)[];
-    columnHeaderRefs: React.RefObject<HTMLTableHeaderCellElement>[];
+    columnsStyles: React.CSSProperties[];
+    columnHeaderRefs: React.RefObject<HTMLTableCellElement>[];
```
Comment (amje) [inline src/components/Table/Table.tsx:22]: Why narrowing down the type?

File: src/components/Table/Table.tsx [hunk 3]
```diff
@@ -428,17 +428,21 @@ export class Table<I extends TableDataItem = Record<string, string>> extends Rea
...
+    private updateColumnStyles() {
+        const columnsWidth = this.state.columnHeaderRefs.map((ref) =>
+            ref.current === null ? undefined : ref.current.getBoundingClientRect().width,
+        );
+
+        const columnsStyles = this.props.columns.map((_, index) =>
+            this.getColumnStyles(index, columnsWidth),
+        );
+
+        this.setState({columnsStyles});
```
Comment (amje) [inline src/components/Table/Table.tsx:440]: Better use function version of setState (`setState((prevState) => newState)`) if you need previous state to build a new state

File: src/components/Table/Table.tsx (hunk 2 above)
Comment (krozhkov) [reply src/components/Table/Table.tsx:22]: HTMLTableHeaderCellElement is deprecated in typescript 4.5.4 definitions.

File: src/components/Table/Table.tsx (hunk 1 above)
Comment (krozhkov) [reply src/components/Table/Table.tsx:104]: I can add an index as optional parameter and if it is not present fall back to the previous logic.

File: src/components/Table/Table.tsx [hunk 4]
```diff
@@ -101,9 +101,9 @@ export class Table<I extends TableDataItem = Record<string, string>> extends Rea
     };
 
     // Static methods may be used by HOCs
-    static getRowId<I extends TableDataItem>(props: TableProps<I>, item: I) {
-        const {data, getRowId} = props;
-        const index = data.indexOf(item);
+    static getRowId<I extends TableDataItem>(props: TableProps<I>, item: I, rowIndex?: number) {
+        const {getRowId} = props;
+        const index = rowIndex ?? props.data.indexOf(item);
```
Comment (amje) [inline src/components/Table/Table.tsx:106]: let's keep vars as they was: `const {data, getRowId} = props;`

File: src/components/Table/Table.tsx (hunk 1 above)
Comment (amje) [reply src/components/Table/Table.tsx]: I think so, in the next major we might want to make the parameter required

**Review by amje — APPROVED** [review]

**Pull request merged**